docs/design/variadics.md

```diff
@@ -0,0 +1,947 @@
+# Variadics
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Basics](#basics)
+    -   [Overview](#overview)
+    -   [Packs and each-names](#packs-and-each-names)
+    -   [Pack expansions](#pack-expansions)
+        -   [Pack expansion expressions and statements](#pack-expansion-expressions-and-statements)
+        -   [Pack expansion patterns](#pack-expansion-patterns)
+    -   [Additional examples](#additional-examples)
+-   [Execution Semantics](#execution-semantics)
+    -   [Expressions and statements](#expressions-and-statements)
+    -   [Pattern matching](#pattern-matching)
+-   [Typechecking](#typechecking)
+    -   [Tuples, packs, segments, and shapes](#tuples-packs-segments-and-shapes)
+    -   [Iterative typechecking of pack expansions](#iterative-typechecking-of-pack-expansions)
+    -   [Typechecking patterns](#typechecking-patterns)
+    -   [Typechecking pattern matches](#typechecking-pattern-matches)
+-   [Appendix: Type system formalism](#appendix-type-system-formalism)
+    -   [Explicit deduced arities](#explicit-deduced-arities)
+    -   [Typing and shaping rules](#typing-and-shaping-rules)
+    -   [Reduction rules](#reduction-rules)
+    -   [Equivalence, equality, and convertibility](#equivalence-equality-and-convertibility)
+    -   [Pattern match typechecking algorithm](#pattern-match-typechecking-algorithm)
+        -   [Canonicalization algorithm](#canonicalization-algorithm)
+-   [Alternatives considered](#alternatives-considered)
+-   [References](#references)
+
+<!-- tocstop -->
+
+## Basics
+
+### Overview
+
+A "pack expansion" is a syntactic unit beginning with `...`, which is a kind of
+compile-time loop over sequences called "packs". Packs are initialized and
```

Comment:
I think changing to talking about runtime loops would make me less happy, because of the cases you mentioned where additional handwaving is required, though perhaps that is more consistent with the exposition here. We can only really give a loop at runtime in the case where the pack is "sufficiently" homogeneous (in similar cases to when we could avoid monomorphization), and that seems to be enough of an edge case that it's probably misleading to center our description around it, even if we do want loop-like lowering for the sufficiently homogeneous cases.